File: 169-MajorityElement.py
from typing import List

'''
My solution
One pass with hashmap
O(n) time, O(n/2) memory
'''
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        dic = {}

        if len(nums) == 1:
            return nums[0]
        for num in nums:
            if num in dic:
                dic[num] = dic[num] + 1
                if dic[num] >= len(nums) / 2:
                    return num
            else:
                dic[num] = 1

'''
Boyer-Moore algorithm copied from leetcode
O(n) time, O(1) memory
'''
class Solution2:
    def majorityElement(self, nums):
        count = 0
        candidate = None

        for num in nums:
            if count == 0:
                candidate = num
            count += 1 if num == candidate else -1

        return candidate

File: 169-MajorityElement.js
/*My Solution

Linear search with hash table
O(n) time, O(n) memory

Other solutions including pre-sorting first O(nlogn) time O(1) memory

Optimal solution is Boyer-Moore Voting Algorithm
O(n) time O(1) memory
*/

/**
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function(nums) {
    let map = {};
    let n = nums.length
    for(let i = 0; i < n;i++){
        let num = nums[i]
        if(map[num]!== undefined){
            map[num] = map[num] + 1;
            if(map[num] > n/2){
                return num
            }
        }
        else{
            map[num] = 1;
        }
    }
    return nums[0];
};

let nums = [2, 1, 1, 1, 3];

console.log(majorityElement(nums));